File: app/src/main/java/com/example/whattowatch/ui/adaptor/EndlessScrollListner.java
package com.example.whattowatch.ui.adaptor;

import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public abstract class EndlessScrollListner extends RecyclerView.OnScrollListener {


    LinearLayoutManager manager;
    int currentItems;
    int totalItems;
    int scrolledItems;
    int page = 1;
    boolean loading = false;



    @Override
    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);
        currentItems = manager.getChildCount();
        totalItems = manager.getItemCount();
        scrolledItems = manager.findFirstVisibleItemPosition();
        scrolledItems = manager.findFirstCompletelyVisibleItemPosition();
        if( !loading && currentItems + scrolledItems == totalItems){

            loading = true;
            page++;
            loadMore(page);
        }

    }

    public abstract void loadMore(int page);

    public void setLoading(boolean loading){
        this.loading = loading;
    }

    public EndlessScrollListner(LinearLayoutManager manager) {
        this.manager = manager;
    }
}

File: app/src/main/java/com/example/whattowatch/ui/adaptor/EndlessScrollListner.kt
package com.example.whattowatch.ui.adaptor

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

abstract class EndlessScrollListner(var manager: LinearLayoutManager) : RecyclerView.OnScrollListener() {
    var currentItems = 0
    var totalItems = 0
    var scrolledItems = 0
    var page = 1
    var loading = false
    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        currentItems = manager.childCount
        totalItems = manager.itemCount
        scrolledItems = manager.findFirstVisibleItemPosition()
        scrolledItems = manager.findFirstCompletelyVisibleItemPosition()
        if (!loading && currentItems + scrolledItems == totalItems) {
            loading = true
            page++
            loadMore(page)
        }
    }

    abstract fun loadMore(page: Int)
    fun setLoading(loading: Boolean) {
        this.loading = loading
    }

}